enhancements/security/x509-trust.md

```diff
@@ -0,0 +1,174 @@
+
+## Proposal
+
+### Cluster-scoped Default Trust
+
+There will be a ConfigMap (FIXME: or should this be a new CRD under openshift/api?) named `default-ca-bundle` in the `openshift-config` namespace.
+The ConfigMap will have the following keys:
+
+* `ca-bundle.crt`, containing a PEM-encoded X.509 certificate bundle.
+* `composition`, configuring how the bundle in `ca-bundle.crt` is composed with the component's default trust store.
+    Valid values are:
+    * `append`, in which case components should use the union of the `ca-bundle.crt` and their default trust store.
+    * `override`, in which case components should use only `ca-bundle.crt` and ignore their default trust store.
+
+### Per-component Trust
+
+Each component will define its own well-known ConfigMap(s) in its own namespace, with the same `ca-bundle.crt` and `composition` keys as [the cluster-scoped ConfigMap](#cluster-scoped-default-trust).
+For example, the registry operator uses [`trusted-ca`][registry-configmap-name] in [the `openshift-image-registry` namespace][registry-configmap-namespace].
+Components that have distinct trust domains may define multiple ConfigMaps for each domain.
+For example, there may be one ConfigMap for proxy/egress TLS, and another ConfigMap for [S/MIME][smime] verification.
+To populate that trust bundle, the cluster administrator can set labels on the ConfigMap(s).
+For example, [the config.openshift.io/inject-trusted-cabundle` label](../proxy/global-cluster-egress-proxy.md#implementation-detailsnotesconstraints-optional) asks for the current proxy trust bundle.
+
+The cluster-version operator (CVO) [merges ConfigMaps][cluster-version-operator-EnsureConfigMap] by [clobbering any manifest-defined data][cluster-version-operator-mergeMap], [labels, and annotations][cluster-version-operator-EnsureObjectMeta].
+Data keys, labels, and annotations not defined in the manifest are ignored.
+This means that components which decide to provide their per-component ConfigMaps via the CVO should not set labels like `config.openshift.io/inject-trusted-cabundle` unless they also set [the `release.openshift.io/create-only` annotation to `true`][cluster-version-operator-create-only], to allow cluster administrators to manipulate the per-component ConfigMap without the CVO stomping on their updates.
+For example, the registry operator [sets `create-only`][registry-configmap-create-only].
+
+#### Universal Base Image Containers
+
+FIXME: UBI-specific docs for this [RHEL behavior](rhel-cert-injection) that show that you don't need to install some non-UBI RHEL stuff to make it work.
+
+Components can mount `ca-bundle.crt` from their ConfigMap(s) under `/usr/share/pki/ca-trust-source/anchors/`.
+Components are then responsible for running `update-ca-trust` (FIXME: do we need `extract` or not?) at start.
+They must also run `update-ca-trust` periodically thereafter to pull in any [updates from the mounted ConfigMap][mounted-configmap-updates].
```

Comment:
running update-ca-trust requires root. not all containers can run as root.

Reply:
> not all containers can run as root.

What's the alternative?  Holding your trust under `~`?  Or do you have to union keys in Kube-space and use the single-bundle `/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem` workaround?

Reply:
the latter is more or less what we did for proxy.

or you need to load your CAs from additional locations, if you have that flexibility in your transport libs.

Reply:
What we really want of course is user namespaces so that "as root" is really only in the container.  xref https://lwn.net/Articles/796700/